### PR TITLE
Split-checkers support produce multiple split keys

### PR DESCRIPTION
**Pull request by Connor1996**
## What have you changed? (mandatory)
For batch split PR, we just make it possible for split-checker to return multiple keys. But actually the implementation of split-checks doesn't change which is still producing only one key. This PR is support to produce multiple split keys.  

## What are the type of the changes? (mandatory)
- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
unit-test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
New config

### CODE REVIEW

File: src/raftstore/coprocessor/split_check/keys.rs [hunk 1]
```diff
@@ -20,37 +22,48 @@ use super::super::{Coprocessor, KeyEntry, ObserverContext, SplitCheckObserver, S
 use super::Host;
 
 pub struct Checker {
-    max_keys: u64,
-    split_keys: u64,
-    current_keys: u64,
-    split_key: Option<Vec<u8>>,
+    max_keys_count: u64,
```
Comment (zhangjinpeng1987) [inline src/raftstore/coprocessor/split_check/keys.rs]: s/split_keys_count/split_threshold

File: tests/raftstore_cases/test_split_region.rs [hunk 2]
```diff
@@ -267,10 +267,10 @@ fn test_auto_split_region<T: Simulator>(cluster: &mut Cluster<T>) {
             Ok(true)
         })
         .expect("");
-    assert!(size <= REGION_SPLIT_SIZE);
-    // although size may be smaller than REGION_SPLIT_SIZE, but the diff should
+    assert!(size >= REGION_SPLIT_SIZE);
```
Comment (Connor1996) [inline tests/raftstore_cases/test_split_region.rs]: for current implementation, the region size is a little bit larger than REGION_SPLIT_SIZE, and before is a little bit smaller than REGION_SPLIT_SIZE. Is that alright? @zhangjinpeng1987 @overvenus

File: src/raftstore/coprocessor/split_check/size.rs [hunk 3]
```diff
@@ -23,35 +25,45 @@ pub struct Checker {
...
-            self.split_key = Some(entry.key().to_vec());
+        let mut over_limit = self.split_keys.len() as u64 >= self.batch_split_limit;
+        if self.current_size >= self.split_size && !over_limit {
+            self.split_keys.push(keys::origin_key(entry.key()).to_vec());
+            self.current_size = 0;
+            over_limit = self.split_keys.len() as u64 >= self.batch_split_limit;
         }
-        // should consider max_size may equal to split_size
-        self.current_size > self.max_size
+        self.current_size += entry.entry_size() as u64;
```
Comment (zhangjinpeng1987) [inline src/raftstore/coprocessor/split_check/size.rs]: Please move L52 before L46.

File: tests/raftstore_cases/test_split_region.rs (hunk 2 above)
Comment (zhangjinpeng1987) [reply tests/raftstore_cases/test_split_region.rs]: To keep consistency, in `on_kv` function, you should accumulate `current_size` first.

File: src/raftstore/coprocessor/split_check/keys.rs [hunk 4]
```diff
@@ -20,37 +22,49 @@ use super::super::{Coprocessor, KeyEntry, ObserverContext, SplitCheckObserver, S
...
-            self.split_key = Some(keys::origin_key(key.key()).to_vec());
+
+        let mut over_limit = self.split_keys.len() as u64 >= self.batch_split_limit;
+        if self.current_count >= self.split_threshold && !over_limit {
+            self.split_keys.push(keys::origin_key(key.key()).to_vec());
+            self.current_count = 0;
+            over_limit = self.split_keys.len() as u64 >= self.batch_split_limit;
         }
-        self.current_keys > self.max_keys
+        self.current_count += 1;
```
Comment (zhangjinpeng1987) [inline src/raftstore/coprocessor/split_check/keys.rs]: Please move L56 before L50.

**Comment by zhangjinpeng1987** [comment]
@BusyJay @overvenus PTAL

File: src/raftstore/coprocessor/config.rs [hunk 5]
```diff
@@ -22,14 +22,19 @@ pub struct Config {
...
 
-    /// When region [a, b) size meets region_max_size, it will be split
-    /// into two region into [a, c), [c, b). And the size of [a, c) will
-    /// be region_split_size (or a little bit smaller).
+    /// For once split check, there are several split_key produced for batch.
+    /// batch_split_limit limits the number of produced split-key for one batch.
+    pub batch_split_limit: u64,
+
+    /// When region [a,b) size meets region_max_size, it will be split into
+    /// several regions [a,c), [c,d), [d,e), [e,b). And the size of [a,c),
```
Comment (overvenus) [inline src/raftstore/coprocessor/config.rs]: [e,f)

File: src/raftstore/coprocessor/split_check/keys.rs [hunk 6]
```diff
@@ -20,37 +22,54 @@ use super::super::{Coprocessor, KeyEntry, ObserverContext, SplitCheckObserver, S
...
+    pub fn new(max_keys_count: u64, split_threshold: u64, batch_split_limit: u64) -> Checker {
         Checker {
-            max_keys,
-            split_keys,
-            current_keys: 0,
-            split_key: None,
+            max_keys_count,
+            split_threshold,
+            current_count: 0,
+            split_keys: Vec::with_capacity(1),
```
Comment (overvenus) [inline src/raftstore/coprocessor/split_check/keys.rs]: Why not batch split limit?

Comment (Connor1996) [reply src/raftstore/coprocessor/split_check/keys.rs]: for most cases, there is only one split key produced.

Comment (Connor1996) [reply src/raftstore/coprocessor/split_check/keys.rs]: for most cases, there is only one split key produced.

**Pull request merged**